### ISSUE TITLE
ignoreAdd does not ignore

### ISSUE THREAD

**Issue by warpech** [Palindrom/Palindrom#82]
@malx122 showed me that ignoreAdd does not ignore properly. Possible clues:
- the first observation comes a an `replace` property, not `add`

This is more likely to be caused by a bug in dirty checking

**Comment by warpech** [comment]
I know what is going on. 

When PuppetJS is applying external patch, Polymer starts it's own notifications. In @malx122 case, this means adding a new property `_collapsed` with a value `true`.

This all happens in period which is **unobserved** in PuppetJS (line https://github.com/PuppetJs/PuppetJs/blob/59f9f1efd1b714e157c9d501db8740c4fbae3dbc/src/puppet.js#L636). Because of that, the newly created property is not observed immediately, which would add it to the `ignoreCache` list.

Now, when PuppetJS resumes observing, JSON-Patch recreates the mirror by cloning the original object again (line https://github.com/Starcounter-Jack/JSON-Patch/blob/6a339cc0c4ffb296e8e816f151530f1f6c2350cc/src/json-patch-duplex.js#L287).

Now I have two possible solutions:

1. Create new methods in JSON-Patch-duplex: `pauseObserving`, `resumeObserving`
2. Remove PuppetJS `ignoreAdd$` altogether, and force users to use `Object.defineProperty` to declare unobervable properties as non-enumerable.

Actually, I would prefer solution 2

**Comment by warpech** [comment]
In point 2, I sugest this:

```js
Object.defineProperty(this.activity, '_collapsed', { value: true, enumerable: true, writable: true });
```

instead of this:

```js
this.activity._collapsed = true;
```

It makes `ignoreAdd` not needed

**Comment by warpech** [comment]
cc @tomalec, @miyconst - WDYT?

**Comment by tomalec** [comment]
2. LGTM
(in your example you probably ment `enumerable: false`

**Comment by warpech** [comment]
> (in your example you probably meant `enumerable: false`)

Sure! Thanks, fixed

**Comment by miyconst** [comment]
I also vote for the second solution, but my concern is Polymer observation and bindings. Does it work with `enumerable: false` properties?

**Comment by malx122** [comment]
Ok, we need to do something fast here since we need to modify puppetjs for every single installation now which is not ok.

we change line 511 from:
```
if (op === 'add' && pattern.test(path))
```
to:
```
if (op === 'add' || pattern.test(path))
```

**Comment by warpech** [comment]
@malx122 Did you try with the solution (2) proposed above?

**Comment by malx122** [comment]
Nope, didn't try. A little afraid about how many places we need to change (and to miss a couple of places)

**Comment by warpech** [comment]
> Nope, didn't try. A little afraid about how many places we need to change (and to miss a couple of places)

I can add a warning when `ignoreAdd` is used, so that you will be notified when you use it.

**Comment by warpech** [comment]
Meanwhile, maybe we can find a way to get rid of `ignoreCache`. Could you check if the following line works OK for you?

```js
if (/*op === 'add' ||*/ pattern.test(path))
```

**Comment by malx122** [comment]
You have to explain Object.defineProperty a little more. Where does it comes from. Do I need to use it every time I change the value or only first time. Is it part of your object observer shim that I'm not a fan of to begin with ;)?

**Comment by malx122** [comment]
That line does not work, but it might be a bug from our side. (We get some random adds that not part of the ignorepattern and that should not have been "add"s, but it happens so seldom that I don't know how to fight it. Still it happens on our client servers and its not allowed to...)

**Comment by warpech** [comment]
[`Object.defineProperty`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty) is a standard JavaScript method that allows to add properties to JS objects with special behavior:

- make a property non-enumerable (that's what we need)
- make a property readonly
- define a setter or getter function

**Comment by malx122** [comment]
Ok, so it does not enumerate on your object observer shim then?
Will it still work within all shims and stuff relating to Polymer on different platforms then?

**Comment by warpech** [comment]
You need to use it only the first time after the object was created (and also after the object was replaced with another object).

So actually this might work: 

```js
function ensureNonEnumerable(obj, prop) {
  if(!obj.hasOwnProperty(prop)) {
    Object.defineProperty(obj, prop, { enumerable: false, writable: true });
  }
  else if(obj.propertyIsEnumerable(prop)) {
    throw new Error("Someone already made the property " + prop + " enumerable");
  }
}
```

Then you need to replace each line that changes the ignored property:

```js
this.activity._collapsed = true;
```

With:

```js
ensureNonEnumerable(this.activity, "_collapsed");
this.activity._collapsed = true;
```

**Comment by warpech** [comment]
> Ok, so it does not enumerate on your object observer shim then?
Will it still work within all shims and stuff relating to Polymer on different platforms then?

We don't have an object observer shim since few months. We do a recursive loop on the `Object.keys` of the object, which will only see the enumerable keys of that object.

> Will it still work within all shims and stuff relating to Polymer on different platforms then?

That is what @miyconst also asked ("*my concern is Polymer observation and bindings. Does it work with enumerable: false properties?*"). 

Basically, you need to check. If it works in one browser, it should work in all of them.

**Comment by warpech** [comment]
>> Meanwhile, maybe we can find a way to get rid of ignoreCache. Could you check if the following line works OK for you?

>>  if (/*op === 'add' ||*/ pattern.test(path))

> That line does not work, but it might be a bug from our side. (We get some random adds that not part of the ignorepattern and that should not have been "add"s, but it happens so seldom that I don't know how to fight it. Still it happens on our client servers and its not allowed to...)

On the second thought that would not work because it changes the meaning of `ignoreAdd`. 

For example `ignoreAdd = /.*/i` means ignore **adding** of properties of any name. Not ignore **adding, replacing, deleting** of properties of any name.

Please check `Object.defineProperty`, because that is our best shot.

**Comment by malx122** [comment]
> We don't have an object observer shim since few months. We do a recursive loop on the Object.keys of the object, which will only see the enumerable keys of that object.

That is kind of a shim/replacement by itself, isn't it?

> Basically, you need to check. If it works in one browser, it should work in all of them.

Nope, since Polymer uses shims ant other stuffs by itself to cover some missing features in some browser, I'm worried that this might not be the case if you start removing standard things like enumerable properties.

**Comment by warpech** [comment]
I created a JSFiddle that tests non-enumerable property with Polymer. It works OK in Chrome, Firefox and Edge. https://jsfiddle.net/warpech/3rjfdxt6/

**Comment by warpech** [comment]
> That is kind of a shim/replacement by itself, isn't it?

Yes, it is a replacement and not a shim.

> Nope, since Polymer uses shims ant other stuffs by itself to cover some missing features in some browser, I'm worried that this might not be the case if you start removing standard things like enumerable properties.

The only shim that Polymer uses is this: https://github.com/webcomponents/webcomponentsjs. It shims 5 different APIs, but none of them is used for notification about data changes. The data change notification protocol is something close to a pub/sub mechanism, implemented just in Polymer and explained here: https://www.polymer-project.org/1.0/docs/devguide/properties.html. Where I am going with this - If Polymer has problems to work with non-enumerable properties, the problem will be evident in all the browsers, including Chrome.

**Comment by alshakero** [comment]
Closing is rendered irrelevant in Palindrom 3.0.0.
